README.md

```diff
@@ -1,4 +1,208 @@
+
+Implementation is based on [Kotlin](https://kotlinlang.org/) [coroutines](https://github.com/Kotlin/kotlinx.coroutines)
+and inspired by both [RxRedux](https://github.com/freeletics/rxredux) library 
+and [Coroutines in practice by Roman Elizarov](https://www.youtube.com/watch?v=a3agLJQ6vt8) KotlinConf2018 talk.
+
+### Basic concept
+
+Imagine, that we need to develop Calculator app that allows to do only following simple actions:
+```kotlin
+sealed class CalculatorAction {
+    data class Add(val value: Int) : CalculatorAction()
+    data class Deduct(val value: Int) : CalculatorAction()
+    data class Multiply(val value: Int) : CalculatorAction()
+    data class Divide(val value: Int) : CalculatorAction()
+}
+```
+
+With CoRedux you can create a store that will be a single source of current Calculator state 
+and will update it on each new incoming `CalculatorAction`:
+```kotlin
+val store = GlobalScope.createStore<Int, CalculatorAction>(
+    name = "Calculator",
+    initialState = 0,
+    reducer = { currentState, newAction ->
+        when (newAction) {
+            is CalculatorAction.Add -> currentState + newAction.value
+            is CalculatorAction.Deduct -> currentState - newAction.value
+            is CalculatorAction.Multiply -> currentState * newAction.value
+            is CalculatorAction.Divide -> currentState / newAction.value
+        }
+    }
+)
+```
+
+Where `reducer` is special function that takes current state, new action that wants to
+modify the state (_intention_) and returns result of state modification, that becomes new current state.
+
+To start receiving current state updates, Calculator UI implementation needs to subscribe to `store` instance:
+```kotlin
+store.subscribe { state -> updateUI(state) }
+```
+By default, when first subscription happens, `store` will emit `initialState` to it.
+
+On each new Calculator UI _intention_, it just need to send it (_dispatch_) as an action to `store` instance:
+```kotlin
+store.dispatch(CalculatorAction.Add(10)) 
+// Current state will become 10
+store.dispatch(CalculatorAction.Deduct(1)) 
+// Current state will become 9
+store.dispatch(CalculatorAction.Add(1)) 
+// Current state will become 10
+store.dispatch(CalculatorAction.Divide(10)) 
+// Current state will become 1
+```
+
+All actions will be processed in a serialized order 
+and on each incoming action `reducer` function will be called to update current state.
+
+### Side effects
+
+Side effect is an interface that has one function - it receives incoming 
+actions, can get current store state at any time and may emit outgoing actions.
+**So basically it is actions in and actions out.** 
+
+**Actions in** to side effects are emitted by store after calling `reducer` function. 
+
+**Actions out** are consumed back again by store and trigger calling `reducer` function.
+
+Basic usage for side effects is to perform some job on incoming action, that may take time (I/O operation) 
```

Comment:
> Side effects are used to perform additional Job as a reaction on a certain `Action` for example making an http request, I/O operations, writing to database and so on. Since they run in a `Job` they can run async. Each Job is scoped to the context of the `store`.